### PR TITLE
[Metric tensor tape] Covariance matrix support

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

The metric tensor is the last main feature not supported in tape mode. This is PR 1 of two PRs, focused on adding support for the metric tensor to the tape. In this PR, we add the ability to compute the covariance matrix of a circuit ansatz with respect to a list of commuting observables, in a differentiable manner.

**Description of the Change:**

* Adds the new function `qml.math.cov_matrix()`. This function accepts a list of commuting observables, and the probability distribution in the shared observable eigenbasis after the application of an ansatz. It uses these to construct the covariance matrix in a *framework independent* manner, such that the output covariance matrix is autodifferentiable.

  For example, consider the following ansatz and observable list:

  ```python
  obs_list = [qml.PauliX(0) @ qml.PauliZ(1), qml.PauliY(2)]
  ansatz = qml.templates.StronglyEntanglingLayers
  ```

  We can construct a QNode to output the probability distribution in the shared eigenbasis of the observables:

  ```python
  dev = qml.device("default.qubit", wires=3)
  
  @qml.qnode(dev, interface="autograd")
  def circuit(weights):
      ansatz(weights, wires=[0, 1, 2])
      # rotate into the basis of the observables
      for o in obs_list:
          o.diagonalizing_gates()
      return qml.probs(wires=[0, 1, 2])
  ````

  We can now compute the covariance matrix:
  ```pycon
  >>> weights = qml.init.strong_ent_layers_normal(n_layers=2, n_wires=3)
  >>> cov = qml.math.cov_matrix(circuit(weights), obs_list)
  >>> cov
  array([[0.98707611, 0.03665537],
         [0.03665537, 0.99998377]])
  ```

  Autodifferentiation is fully supported using all interfaces:

  ```pycon
  >>> cost_fn = lambda weights: qml.math.cov_matrix(circuit(weights), obs_list)[0, 1]
  >>> qml.grad(cost_fn)(weights)[0]
  array([[[ 4.94240914e-17, -2.33786398e-01, -1.54193959e-01],
          [-3.05414996e-17,  8.40072236e-04,  5.57884080e-04],
          [ 3.01859411e-17,  8.60411436e-03,  6.15745204e-04]],

         [[ 6.80309533e-04, -1.23162742e-03,  1.08729813e-03],
          [-1.53863193e-01, -1.38700657e-02, -1.36243323e-01],
          [-1.54665054e-01, -1.89018172e-02, -1.56415558e-01]]])
  ```

* In order to support this new function, the following low-level tensor functions were added to `qml.math`:
  - `diag` (analogous to `np.diag`)
  - `flatten`
  - `marginal_prob` (marginalize a probability distribution over given axes)
  - `reshape`
  - `scatter_element_add` (a functional equivalent of `tensor[idx] += value`)

**Benefits:**

* The ability to compute the covariance matrix is crucial to computing the metric tensor on hardware. The block diagonal approximation to the metric tensor is simply a sequence of covariance matrix computations, one per layer of the original QNode, executed in batch.

* Since the implementation of the covariance matrix here is autodifferentiable, the metric tensor computation in the follow up PR will also be differentiable.

* I have added `cov_matrix` to `qml.math` because it is a low-level function of the form `tensor -> tensor`. It is _not_ a tape transform.

**Possible Drawbacks:** n/a

**Related GitHub Issues:** n/a

### CODE REVIEW

File: pennylane/math/autograd_box.py [hunk 1]
```diff
@@ -78,6 +80,15 @@ def numpy(self):
...
         return self.data.requires_grad
 
+    @wrap_output
+    def scatter_element_add(self, index, value):
+        size = self.data.size
+        flat_index = np.ravel_multi_index(index, self.shape)
+        t = [0] * size
+        t[flat_index] = value
+        self.data = self.data + np.array(t).reshape(self.shape)
+        return self.data
```
Comment (josh146) [inline pennylane/math/autograd_box.py:90]: The convoluted approach here is because autograd does not support array assignment, nor does numpy provide a functional approach to scattering :(

If anyone has any insights on a better way of doing this (while preserving differentiabilty in autograd), would appreciate it!

File: tests/math/test_functions.py [hunk 2]
```diff
@@ -958,3 +958,157 @@ def test_squeeze(t):
...
+
+    @staticmethod
+    def expected_grad(weights):
+        """Analytic covariance matrix gradient for ansatz and obs_list"""
+        a, b, c = weights
+        return np.array([
+            np.sin(a) * np.sin(b) ** 2 * np.sin(c),
+            -2 * np.cos(a) * np.cos(b) * np.sin(b) * np.sin(c),
+            -np.cos(a) * np.cos(c) * np.sin(b) ** 2
+        ])
```
Comment (josh146) [inline tests/math/test_functions.py:993]: I computed the covariance matrix of this ansatz+obs_list by hand, and differentiated it by hand, to ensure that all interfaces were producing the correct result.

File: pennylane/math/fn.py [hunk 3]
```diff
@@ -339,6 +492,39 @@ def get_interface(tensor):
     return TensorBox(tensor).interface
 
 
+def marginal_prob(prob, axis):
+    """Compute the marginal probability given joint probabilities of the full system.
```
Comment (mariaschuld) [inline pennylane/math/fn.py:496]: Hm, the implementation makes sense for quantum probabilities, but the function is named as if it is agnostic of how the probability came about....rename to "marginal_quantum_prob"? Ok, that looks ugly...Or maybe this:
```suggestion
    """Compute the marginal probability given a joint probability distribution expressed as a tensor.
    Each random variable corresponds to a dimension. 
    
    If the distribution arises from a quantum circuit measured in computational basis, each dimension 
    corresponds to a wire. For example, for a 2-qubit quantum circuit `prob[0, 1]` is the probability of measuring the 
    first qubit in state 0 and the second in state 1. 
```

File: pennylane/math/fn.py [hunk 4]
```diff
@@ -339,6 +492,39 @@ def get_interface(tensor):
...
+    **Example**
+
+    >>> x = tf.Variable([1, 0, 0, 1.], dtype=tf.float64) / np.sqrt(2)
+    >>> marginal_prob(x, axis=[0, 1])
+    <tf.Tensor: shape=(4,), dtype=float64, numpy=array([0.70710678, 0.        , 0.        , 0.70710678])>
+    >>> marginal_prob(x, axis=[0])
+    <tf.Tensor: shape=(2,), dtype=float64, numpy=array([0.70710678, 0.70710678])>
+    """
+    prob = flatten(prob)
+    num_wires = int(np.log2(len(prob)))
```
Comment (mariaschuld) [inline pennylane/math/fn.py:517]: Change to `num_vars`?

File: pennylane/math/fn.py [hunk 5]
```diff
@@ -339,6 +492,39 @@ def get_interface(tensor):
     return TensorBox(tensor).interface
 
 
+def marginal_prob(prob, axis):
+    """Compute the marginal probability given joint probabilities of the full system.
+
+    Args:
+        prob (tensor_like): 1D tensor of probabilities. This tensor should of size
+            ``(2**N,)`` for some integer value ``N``.
+        axis (list[int]): the axis for which to calculate the marginal
```
Comment (mariaschuld) [inline pennylane/math/fn.py:501]:
```suggestion
        axes (list[int]): the axes for which to calculate the marginal
```
not sure if numpy convention is `axis`?

File: pennylane/math/fn.py [hunk 6]
```diff
@@ -339,6 +492,39 @@ def get_interface(tensor):
...
+    <tf.Tensor: shape=(2,), dtype=float64, numpy=array([0.70710678, 0.70710678])>
+    """
+    prob = flatten(prob)
+    num_wires = int(np.log2(len(prob)))
+
+    if num_wires == len(axis):
+        return prob
+
+    inactive_wires = tuple(set(range(num_wires)) - set(axis))
+    prob = reshape(prob, [2] * num_wires)
```
Comment (mariaschuld) [inline pennylane/math/fn.py:523]: Is this reshape to avoid copying? Or is something more going on?

File: pennylane/math/torch_box.py [hunk 7]
```diff
@@ -117,6 +126,12 @@ def numpy(self):
     def requires_grad(self):
         return self.data.requires_grad
 
+    @wrap_output
+    def scatter_element_add(self, index, value):
+        self.data = self.data.clone()
+        self.data[tuple(index)] += value
+        return self.data
```
Comment (mariaschuld) [inline pennylane/math/torch_box.py:133]: So this function is implemented not in-place?

File: tests/math/test_functions.py [hunk 8]
```diff
@@ -958,3 +958,208 @@ def test_squeeze(t):
...
+    def test_sequence(self, a, interface):
+        """Test that a sequence is automatically converted into
+        a diagonal tensor"""
+        t = [0.1, 0.2, a]
+        res = fn.diag(t)
+        assert fn.get_interface(res) == interface
+        assert fn.allclose(res, onp.diag([0.1, 0.2, 0.5]))
+
+    def test_array(self):
+        """Test that sum, called without the axis arguments, returns a scalar"""
```
Comment (mariaschuld) [inline tests/math/test_functions.py:976]: Check docstring here and below?

File: tests/math/test_functions.py (hunk 2 above)
Comment (mariaschuld) [reply tests/math/test_functions.py:1045]: That's commitment!

File: tests/math/test_functions.py [hunk 9]
```diff
@@ -958,3 +958,208 @@ def test_squeeze(t):
...
+            for o in self.obs_list:
+                o.diagonalizing_gates()
+
+            return qml.probs(wires=[0, 1, 2])
+
+        def cov(weights):
+            probs = circuit(weights)
+            return fn.cov_matrix(probs, self.obs_list)
+
+        weights = np.array([0.1, 0.2, 0.3])
```
Comment (mariaschuld) [inline tests/math/test_functions.py:1067]: Don't you want to make this test data, maybe using some edge cases?

**Review by mariaschuld — APPROVED** [review]
Awesome @josh146 !

I approve in the interest of time, but check out the comments, there may be one or two minor fixes needed.

File: pennylane/math/fn.py [hunk 10]
```diff
@@ -244,6 +246,98 @@ def concatenate(values, axis=0):
...
 
 
+def cov_matrix(prob, obs):
+    """Calculate the covariance matrix of a list of commuting observables, given
+    the joint probability distribution of the system in the shared eigenbasis.
+
+    .. note::
+        This method only works for **commuting observables.**
+        The probability distribution must be rotated into the shared
+        eigenbasis of the list of observables.
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/math/fn.py:256]: Yep, that's better wording :)

File: pennylane/math/fn.py (hunk 5 above)
Comment (josh146) [reply pennylane/math/fn.py:501]: Just checked, seems to be `axis`: https://numpy.org/doc/stable/reference/generated/numpy.sum.html

File: pennylane/math/fn.py (hunk 4 above)
Comment (josh146) [reply pennylane/math/fn.py:517]: What does vars refer to here?

File: pennylane/math/fn.py (hunk 6 above)
Comment (josh146) [reply pennylane/math/fn.py:523]: Oh, we are reshaping from a flat array of size 2**N (as returned from the QNode) to a N-dimensional array :) Makes it easier to simply sum over the axes we no longer care about.

But yes the reshape avoids copying as well.

File: pennylane/math/torch_box.py (hunk 7 above)
Comment (josh146) [reply pennylane/math/torch_box.py:133]: It's a limitation of PyTorch :(


PyTorch allows in-place operations of tensors, *unless* the tensor is a leaf node (that is, a tensor created at the beginning of the computational graph by a user, with the `requires_grad` explicitly set).

But I just realised we can have a check for this using `self.data.is_leaf`, so we only need to clone if that is true.

File: pennylane/math/torch_box.py [hunk 11]
```diff
@@ -117,6 +126,12 @@ def numpy(self):
     def requires_grad(self):
         return self.data.requires_grad
 
+    @wrap_output
+    def scatter_element_add(self, index, value):
+        self.data = self.data.clone()
```
Comment (josh146) [inline pennylane/math/torch_box.py:131]:
```suggestion
        if self.data.is_leaf:
            self.data = self.data.clone()
```

File: pennylane/math/fn.py [hunk 12]
```diff
@@ -244,6 +246,98 @@ def concatenate(values, axis=0):
...
+        p1 = marginal_prob(prob, o1.wires)
+        p2 = marginal_prob(prob, o2.wires)
+        p12 = marginal_prob(prob, o1.wires + o2.wires)
+
+        res = dot(l12, p12) - dot(l1, p1) * dot(l2, p2)
+
+        cov = scatter_element_add(cov, [i, j], res)
+        cov = scatter_element_add(cov, [j, i], res)
+
+    return cov
```
[the remark this answers is not in the document]
Comment (josh146) [reply pennylane/math/fn.py:346]: With `qml.math`, we could really try and push this. Any pre- or post- quantum processing function we add, we should always now try and add in a differentiable manner :D

File: pennylane/math/fn.py [hunk 13]
```diff
@@ -323,13 +326,17 @@ def circuit(weights):
         o1 = obs[i]
         o2 = obs[j]
 
+        o1wires = o1.wires.labels if wires is None else wires.indices(o1.wires)
+        o2wires = o2.wires.labels if wires is None else wires.indices(o2.wires)
+        shared_wires = o1wires + o2wires
```
Comment (mariaschuld) [inline pennylane/math/fn.py:331]: Wait, the sum would be combined wires, not shared ones, right? If you want to make sure here that the obs do not act on the same wires you could do `Wires.all_wires(o1.wires, o2.wires)` and then extract label or indices?

File: tests/math/test_functions.py [hunk 14]
```diff
@@ -1044,6 +1048,37 @@ def expected_grad(weights):
             -np.cos(a) * np.cos(c) * np.sin(b) ** 2
         ])
 
+    def test_weird_wires(self, in_tape_mode, tol):
```
Comment (mariaschuld) [inline tests/math/test_functions.py:1118]: lol. `custom`?

**Review by mariaschuld — APPROVED** [review]
Looks good!

File: pennylane/math/fn.py (hunk 13 above)
Comment (josh146) [reply pennylane/math/fn.py:331]: Ah! The reason I did this was a bit weird. Basically, by the time this variable ends up in low-level processing functions, the duplicated wires no longer matter. So I removed the call to `all_wires`, because of overhead fear 😆 

Including it is definitely more 'clean', but it adds unnecessary overhead...

How about this, as a cleaner way of writing it?

```suggestion
        shared_wires = set(o1wires + o2wires)
```

File: tests/math/test_functions.py (hunk 14 above)
Comment (josh146) [reply tests/math/test_functions.py:1118]: hahaha I believe @johannesjmeyer has a test called `test_weird_wires` in the circuit drawer, so I stole the name from there

Comment (josh146) [reply tests/math/test_functions.py:1118]: or maybe I'm imagining it and made it up

Comment (johannesjmeyer) [reply tests/math/test_functions.py:1118]: Sounds plausible though
